### ISSUE TITLE
Remove configuration aspect from onInitialConfiguration?

### ISSUE THREAD

**Issue by bubba** [alanz/lsp#253]
A lot of servers seem to be using `initializationOptions` in the `initialize` request to pass the workspace configuration in, but see the discussion here https://github.com/microsoft/language-server-protocol/issues/567 and the official recommendation here: 
https://github.com/microsoft/language-server-protocol/commit/545cd6958ace603cd254e8e376e4fb2840bb1223

It looks like they are indeed separate things, and that `initializationOptions` is more for command line arguments etc. However our API currently suggests that they should be treated the same, by providing a callback to extract the workspace configuration from `initializationOptions` and store it in the `resConfig` state. 

```haskell
data InitializeCallbacks config =
  InitializeCallbacks
    { onInitialConfiguration :: InitializeRequest -> Either T.Text config
```

To me it seems like the `resConfig` state should be used to keep track of the most recent `workspace/didChangeConfiguration` notifications exclusively (and we should extend this to be updated whenever a `workspace/configuration` request is made from the server). 
Should we change `onInitialConfiguration` to something more like:

```haskell
data InitializeCallbacks config =
  InitializeCallbacks
    { onInitialize :: InitializeRequest -> LspM config ()
```

to guide server authors down the right path. This `LspM` monad is from the type safe branch in #244.  If authors still need to get the workspace configuration immediately after initialization, then in `LspM` they can make a `workspace/configuration` request to pull it from the client.

**Comment by bubba** [comment]
cc @alanz @wz1000

**Comment by jneira** [comment]
Agree, when working in the config acess in `Rules` i did not see in the protocol docs nothing about two sources of config should be the same (both are `Any` 😜 ) so i think they should be separated in our code. Even if they are the same at the end, for some server implementations.

**Comment by bubba** [comment]
If vscode sends a `didConfigurationChange` on start-up then that should probably the norm that other clients follow too

**Comment by bubba** [comment]
What I have so far looks like:

```haskell
-- | Contains all the callbacks to use for initialized the language server.
-- it is parameterized over a config type variable representing the type for the
-- specific configuration data the language server needs to use.
data InitializeCallbacks config =
  InitializeCallbacks
    { onConfigurationChange :: J.Value -> LspM config (Either T.Text config)
      -- ^ @onConfigurationChange newConfig@ is called whenever the
      -- clients sends a message with a changed client configuration. This
      -- callback should return either the parsed configuration data or an error
      -- indicating what went wrong. The parsed configuration object will be
      -- stored internally and can be accessed via 'config'.
    , onInitialization :: InitializeRequest -> LspM config (Maybe ResponseError)
      -- ^ Called after receiving the initialization request and before returning the response.
      -- This callback will be invoked to offer the language server
      -- implementation the chance to create any processes or start new threads
      -- that may be necesary for the server lifecycle.
      -- It can also return an error in the initialization if necessary.
    }
```

**Comment by alanz** [comment]
see also #210 and #211

**Comment by alanz** [comment]
The intention behind the init is you may be using it for one-time startup stuff. And there is no guarantee that they should be the same, so we should allow them to be different.

That said, command line args to launch the server are probably the best way to actually pass init options.

**Comment by wz1000** [comment]
Fixed in #244
